pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -362,7 +363,12 @@ private void readMoreEntries(Consumer consumer) {
                 }
                 havePendingRead = true;
                 if (consumer.readCompacted()) {
-                    boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+                    boolean readFromEarliest = false;
+                    if (!cursor.isDurable() || cursor.getName().equals(Compactor.COMPACTION_SUBSCRIPTION)
+                            || cursor.getMarkDeletedPosition() == null
+                            || cursor.getMarkDeletedPosition().getEntryId() == -1L) {
+                        readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+                    }
```

Comment:
```suggestion
                    boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId())
                            && (!cursor.isDurable()
                            || cursor.getName().equals(Compactor.COMPACTION_SUBSCRIPTION)
                            || cursor.getMarkDeletedPosition() == null
                            || cursor.getMarkDeletedPosition().getEntryId() == -1L);
```

Merge all conditions into a single expression. And you don't need to check `cursor.xxx()` when `isFirstRead` is false after that.